src/apps/socket/raw.lua

```diff
@@ -1,72 +1,123 @@
-module(...,package.seeall)
+module(..., package.seeall)
 
-local app    = require("core.app")
-local link   = require("core.link")
+local S = require("syscall")
+local h = require("syscall.helpers")
+local bit = require("bit")
+local link = require("core.link")
 local packet = require("core.packet")
-local dev    = require("apps.socket.dev").dev
+local ffi = require("ffi")
+local C = ffi.C
+
+local c, t = S.c, S.types.t
 
 RawSocket = {}
 
 function RawSocket:new (ifname)
    assert(ifname)
-   self.__index = self
-   return setmetatable({dev = dev:new(ifname)}, self)
+   local tp = h.htons(c.ETH_P["ALL"])
+   local sock, err = S.socket(c.AF.PACKET, bit.bor(c.SOCK.RAW, c.SOCK.NONBLOCK), tp)
+   if not sock then return nil, err end
+   local index, err = S.util.if_nametoindex(ifname, sock)
+   if err then
+      S.close(sock)
+      return nil, err
+   end
+   local addr = t.sockaddr_ll{sll_family = c.AF.PACKET, sll_ifindex = index, sll_protocol = tp}
+   local ok, err = S.bind(sock, addr)
+   if not ok then
+      S.close(sock)
+      return nil, err
+   end
+   return setmetatable({sock = sock}, {__index = RawSocket})
 end
 
 function RawSocket:pull ()
    local l = self.output.tx
    if l == nil then return end
-   while not link.full(l) and self.dev:can_receive() do
-      link.transmit(l, self.dev:receive())
+   while not link.full(l) and self:can_receive() do
+      link.transmit(l, self:receive())
    end
 end
 
+function RawSocket:can_receive ()
+   local ok, err = S.select({readfds = {self.sock}}, 0)
+   return not (err or ok.count == 0)
+end
+
+function RawSocket:receive ()
+   local p = packet.allocate()
+   local sz, err = S.read(self.sock, p.data, C.PACKET_PAYLOAD_SIZE)
```

Comment:
I don't know. So far for accessing the `data` and `length` attributes of a packet I simply accessed those attributes directly, but I just saw there's a `packet.data` and `packet.length` functions in `src/core/packet.lua`. The point is that these functions are getters but there's no equivalent setter methods and in the code I need to update the length too. To me it seems more consistent to access these attributes always the same way. 

What's the preferred way of accessing these attributes? I can change it but I don't understand what are the benefits of using these functions.